### PR TITLE
Replace android-cropimage with uCrop

### PR DESCRIPTION
**Pull request by AndyG**
Replaces [android-cropimage](https://github.com/lvillani/android-cropimage) with [uCrop](https://github.com/Yalantis/uCrop) as the cropping solution. This does provide several additional cropping features such as rotation and aspect ratio changes.

Also bumps various build tool versions.

- [ ] Release a beta and test on various devices
- [ ] Verify that we can theme the activity in a client of this library

### CODE REVIEW

File: media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java [hunk 1]
```diff
@@ -177,19 +178,19 @@ public static void startForImageCrop(final Provider provider, final File file, i
      */
     private static void startForImageCrop(final Provider provider, final Uri uri, int outputWidth, int outputHeight, int colorInt, final OnError onError) {
```
Comment (AndyG) [inline media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java]: need to get rid of `colorInt` and provide users ability to provide their own `UCrop.Options`.

File: media-picker/build.gradle [hunk 2]
```diff
@@ -18,15 +18,21 @@ android {
...
-    //noinspection GradleDynamicVersion || File provider compatibility.
-    compile 'com.android.support:support-core-utils:25.+'
+    // File provider compatibility.
+    compile 'com.android.support:support-core-utils:25.3.1'
 
-    //noinspection GradleDynamicVersion || File cropping utility.
-    compile 'me.villani.lorenzo.android:android-cropimage:1.+'
+    // File cropping utility.
+    compile 'com.github.yalantis:ucrop:2.2.1'
+    compile 'com.github.yalantis:ucrop:2.2.1-native'
```
Comment (mrkcsc) [inline media-picker/build.gradle]: An additional 1.5MB seems like a steep price to pay for a small bump in image quality.

AFAIK most upstream consumers (Discord included) are going to already be compressing to a JPEG so not sure if we should be forcing it at this level, perhaps omit it and add a note in the readme that library consumers can choose to include this explicitly if they wish.

File: media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java [hunk 3]
```diff
@@ -158,38 +160,49 @@ public static void startForDocuments(final Provider provider, final OnError onEr
...
+            // Force JPEG format compression now while working out filename strategy.
+            cropOptions.setCompressionFormat(Bitmap.CompressFormat.JPEG);
 
-            intentBuilder.setSourceImage(uri);
-            intentBuilder.setDoFaceDetection(false);
-            intentBuilder.setOutlineCircleColor(colorInt);
-            intentBuilder.setOutlineColor(colorInt);
-            intentBuilder.setScaleUpIfNeeded(true);
+            final Context context = provider.getContext();
+            final Uri destUri = Uri.fromFile(new File(context.getFilesDir(), "cropped.jpeg"));
```
Comment (mrkcsc) [inline media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java]: Should call `createTempImageFileAndPersistUri(provider)` to be consistent with how the files get created, similar to `startForCamera`.

File: media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java [hunk 4]
```diff
@@ -158,38 +160,49 @@ public static void startForDocuments(final Provider provider, final OnError onEr
...
+
+    public static void startForImageCrop(final Provider provider, final Uri uri, int maxOutputWidth, int maxOutputHeight, final OnError onError, final @Nullable UCrop.Options options) {
         try {
-            final Uri captureFileURI = createTempImageFileAndPersistUri(provider);
+            final UCrop.Options cropOptions = options != null
+                ? options
+                : new UCrop.Options();
 
-            final CropImageIntentBuilder intentBuilder = new CropImageIntentBuilder(outputWidth, outputHeight, captureFileURI);
+            // Force JPEG format compression now while working out filename strategy.
```
Comment (mrkcsc) [inline media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java:193]: Can you elaborate on this comment a bit, not sure exactly what it means.

File: media-picker/build.gradle (hunk 2 above)
Comment (AndyG) [reply media-picker/build.gradle]: Yes that would be a better idea. I thought I needed this but actually didn't (was specifying bad params) and never got rid of it.

File: media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java (hunk 4 above)
Comment (AndyG) [reply media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java:193]: This is related to file creation woes I had, see responses to https://github.com/mrkcsc/android-media-picker/pull/21/files#r121267528.

File: media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java (hunk 3 above)
Comment (AndyG) [reply media-picker/src/main/java/com/miguelgaeta/media_picker/MediaPicker.java]: This is where I need some help in this PR. When I use that, even though I am calling the `grantWriteAccessToURI` method, the cropping fails -- I believe during the step when our activity tries to read the result of what the cropping activity wrote to. This is a cheap sidestep of the problem to avoid spending too much time figuring out something that you probably already know the answer to.

tl;dr this is a stopgap til Monday when I can pick your brain a little :)

**Comment by lytefast** [comment]
https://github.com/mrkcsc/android-media-picker/pull/22 should fix the loading issue. However you might need to update code to work with Uris now

File: app/src/main/java/com/miguelgaeta/android_media_picker/AppActivity.java [hunk 5]
```diff
@@ -88,11 +87,11 @@ public void onSuccess(final Uri uri, final String mimeType, RequestType request)
 
                 if (request != RequestType.CROP && MimeType.isImage(mimeType)) {
 
-                    final int paramColor = ContextCompat.getColor(AppActivity.this, android.R.color.black);
+//                    final int paramColor = ContextCompat.getColor(AppActivity.this, android.R.color.black);
```
Comment (mrkcsc) [inline app/src/main/java/com/miguelgaeta/android_media_picker/AppActivity.java]: Unrevert

**Review by mrkcsc — APPROVED** [review]
